articles/external_qos_file.md

```diff
@@ -0,0 +1,402 @@
+---
+layout: default
+title: Configuring QoS from an external file
+permalink: articles/qos_file.html
+abstract:
+  This article describes how to configure QoS from an external file.
+author: '[Ivan Santiago Paunovic](https://github.com/ivanpauno)'
+published: true
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Overview
+
+`ROS 2` added the [Quality of Service (QoS)](qos.md) to publishers, subscriptions, clients and services.
+Up `ROS 2 Foxy`, the `QoS` can only be specified in code.
+To avoid recompiling the source code with patched `QoS`, node implementers have used different mechanisms to configure `QoS` profiles:
+- Command line arguments
+- ROS parameters
+- Combining system default QoS with vendor specific `QoS` profiles.
+
+Instead of having different ways of externally specifying `QoS` for the different endpoints, it would be good to have a standardized way to specific `QoS` profiles.
+
+## QoS file format
+
+DDS uses the concept of QoS profiles ([rti documentation](https://community.rti.com/examples/using-qos-profiles)), in which each profile has a name. When creating an entity (e.g. a `DataWriter`), the passed profile name will be looked up in the qos profiles file being used, and the QoS can be loaded in that way.
+
+ROS 2 can use a simpler approach, leveraging node name uniqueness <sup id="back_node_name_uniqueness">[1](#to_node_name_uniqueness)</sup>.
+
+<b id="to_node_name_uniqueness">1</b> Node name uniqueness is not being enforced up to Foxy, though it is supposed. [↩](#back_node_name_uniqueness)
+
+### Nodes
+
+In an acceptable `QoS` file format, it should be to distinguish `QoS` settings targeted to different nodes.
+Thus, a top level tag to identify the node should be available.
+Example alternatives:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        ...
+    </node>
+<qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        ...
+```
+
+### Endpoint type
+
+It should be possible to distinguish between the different endpoint types in a qos profile file.
+For example:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd">  <!--Relative name-->
+            ...
+        </publisher>
+        <subscription topic_name="/another_ns/asd">  <!--Absolute name-->
+            ...
+        </subscription>
+    </node>
+<qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            qos:
+                ...
+        subscription:
+            topic_name: /another_ns/asd
+            qos:
+                ...
+```
+
+Besides `publisher` and `subscription` tags, `client` and `service` should be allowed.
```

Comment:
DDS QoS profiles are not specified as reader or writer. If we need to distinguish between publisher and subscriber (rather than just saying "topic QoS" or even generic QoS) then the reason should be given.

Reply:
> DDS QoS profiles are not specified as reader or writer. If we need to distinguish between publisher and subscriber (rather than just saying "topic QoS" or even generic QoS) then the reason should be given.

I think they are.
e.g.: deadline QoS policy isn't part of the reader QoS, and it's part of writer QoS.

You also have topic QoS, that become the default for readers and publishers on that topic (and can be modified for a particular publisher/subscription).

---

The idea here wasn't to differentiate the subscription and publisher QoS profiles (which I think should be distinguished).
It was rather have a mechanism to set different QoS settings in a publisher and subscription on the same topic within a node.
e.g.: A publisher with history depth 1 and a subscription with history depth 100 in the same topic is not a strange combination.

There might be the case that you want to repeat the same profile for both.
For that, I think that the proposal should have a way to define a profile in the file, that can later be reused.
The proposal is lacking that feature, but I'm planning to add it.
Does that make sense?

Reply:
Yes, that makes sense.